File: ScaleformUI_Lua/src/Menus/PauseMenus/PauseMenu/Tabs/PlayerListTab.lua
PlayerListTab = {}
PlayerListTab.__index = PlayerListTab
setmetatable(PlayerListTab, { __index = BaseTab })
PlayerListTab.__call = function() return "PlayerListTab" end

---@enum PLT_COLUMNS
PLT_COLUMNS =
{
    SETTINGS = 0,
    PLAYERS = 1,
    MISSION = 2,
    STORE = 3,
    MISSION_DETAILS = 4,
}

---@class PlayerListTab
---@field public Base BaseTab
---@field public LeftItemList PauseMenuItem[]
---@field public Label string
---@field public TextTitle string
---@field public SettingsColumn SettingsListColumn
---@field public PlayersColumn PlayerListColumn
---@field public MissionsColumn MissionListColumn
---@field public StoreColumn StoreListColumn
---@field public Index number
---@field public Focused boolean

---Creates a new PlayerListTab.
---@param name string
---@param color SColor
---@return PlayerListTab
function PlayerListTab.New(name, color)
    local base = BaseTab.New(name, color)
    base._identifier = "Page_Multi"
    base.OnFocusChanged = function(focus)
    end
    base.order = { 0, 0, 0 }
    local meta = setmetatable(base, PlayerListTab)
    meta.Minimap = MinimapPanel.New(meta)
    meta.Minimap.HidePedBlip = true
    return meta
end

function PlayerListTab:SetupLeftColumn(column)
    column.position = 0
    self.LeftColumn = column
    self.LeftColumn.Parent = self
    self.order[1] = column.type
end

function PlayerListTab:SetupCenterColumn(column)
    column.position = 1
    self.CenterColumn = column
    self.CenterColumn.Parent = self
    self.order[2] = column.type
end

function PlayerListTab:SetupRightColumn(column)
    column.position = 2
    self.RightColumn = column
    self.RightColumn.Parent = self
    self.order[3] = column.type
end

function PlayerListTab:SelectColumn(index)
    self:SwitchColumn(index - 1)
end

function PlayerListTab:SwitchColumn(index)
    if index > 2 or self.order[index + 1] == PLT_COLUMNS.MISSION_DETAILS then return end
    local canHideShow = true
    local col = self:GetColumnAtPosition(index)
    if col == nil then
        if index < 2 then
            if index < self.CurrentColumnIndex then
                self:SwitchColumn(index - 1)
            else
                self:SwitchColumn(index + 1)
            end
        end
        return
    end


    -- we don't check for right column because
    -- if right column is players then it means there's no player panel to be shown
    if self.LeftColumn.type == PLT_COLUMNS.PLAYERS then
        if self.LeftColumn:CurrentItem():KeepPanelVisible() then
            canHideShow = false
        end
    elseif self.CenterColumn.type == PLT_COLUMNS.PLAYERS then
        if self.CenterColumn:CurrentItem():KeepPanelVisible() then
            canHideShow = false
        end
    end

    if canHideShow then
        if self.Parent ~= nil and self.Parent:Visible() then
            if col.type == PLT_COLUMNS.PLAYERS then
                if col:CurrentItem().Panel ~= nil then
                    self.RightColumn:ColumnVisible(false)
                end
            else
                -- we check that the columns before and after selected index are players columns
                -- and that its current item has KeepPanelVisible true.. 
                -- if KeepPanelVisible is true, we keep hidden the right column, else we show it.
                local show = true
                local befCol = self:GetColumnAtPosition(index - 1)
                local afterCol = self:GetColumnAtPosition(index + 1)
                if befCol ~= nil and befCol.type == PLT_COLUMNS.PLAYERS and not befCol:CurrentItem():KeepPanelVisible() then
                    befCol:CurrentItem():Dispose()
                    show = false
                end
                if afterCol ~= nil and afterCol.type == PLT_COLUMNS.PLAYERS and not afterCol:CurrentItem():KeepPanelVisible() then
                    afterCol:CurrentItem():Dispose()
                    show = false
                end
                if self.RightColumn ~= nil then
                    self.RightColumn:ColumnVisible(not show)
                end
            end
        end
    else
        if self.RightColumn ~= nil then
            self.RightColumn:ColumnVisible(true)
        end
    end

    if index == 2 and not canHideShow then return end
    self.CurrentColumnIndex = index
    ScaleformUI.Scaleforms._pauseMenu._pause:CallFunction("SET_MENU_LEVEL", self.CurrentColumnIndex + 1)
    ScaleformUI.Scaleforms._pauseMenu._pause:CallFunction("MENU_SHIFT_DEPTH", 0, true, true)
    self.Parent.focusLevel = self.CurrentColumnIndex + 1
    ScaleformUI.Scaleforms._pauseMenu._pause:CallFunction("SET_COLUMN_HIGHLIGHT", col.position, col.index - 1, true, true)
    col.Items[col.index]:Selected(true)
end

function PlayerListTab:StateChange()
    ScaleformUI.Scaleforms._pauseMenu._pause:CallFunction("MENU_STATE", Join("", self.order))
end

function PlayerListTab:GoUp()
    if not self.Focused then return end
    local col = self:CurrentColumn()
    Citizen.CreateThread(function()
        col:GoUp()
    end)
    col:SetColumnScroll(col.index, #col.Items, col.VisibleItems, "", #col.Items < col.VisibleItems)
end

function PlayerListTab:GoDown()
    if not self.Focused then return end
    local col = self:CurrentColumn()
    Citizen.CreateThread(function()
        col:GoDown()
    end)
    col:SetColumnScroll(col.index, #col.Items, col.VisibleItems, "", #col.Items < col.VisibleItems)
end

function PlayerListTab:GoLeft()
    if not self.Focused then return end
    self:CurrentColumn():GoLeft()
end

function PlayerListTab:GoRight()
    if not self.Focused then return end
    self:CurrentColumn():GoRight()
end

function PlayerListTab:Select()
    if not self.Focused then return end
    self:CurrentColumn():Select()
end

function PlayerListTab:GoBack()
    if not self.Focused then return end
    if self.CurrentColumnIndex > 0 then
        self:CurrentColumn():CurrentItem():Selected(false)
        ScaleformUI.Scaleforms._pauseMenu._pause:CallFunction("SET_COLUMN_FOCUS", self.CurrentColumnIndex, false, false,
            false)
        self:SwitchColumn(self.CurrentColumnIndex - 1)
        if self:CurrentColumn() == nil and self.CurrentColumnIndex > 0 then
            self:GoBack()
            return
        end
        self:CurrentColumn():CurrentItem():Selected(true)
        if self:CurrentColumn().type == PLT_COLUMNS.PLAYERS then
            if self:CurrentColumn():CurrentItem().Panel ~= nil then
                local rr = self:GetColumnAtPosition(2)
                if rr ~= nil then
                    rr:ColumnVisible(false)
                end
            else
                local rr = self:GetColumnAtPosition(2)
                if rr ~= nil then
                    rr:ColumnVisible(true)
                end
            end
        end
    end
end

function PlayerListTab:MouseEvent(eventType, context, index)
    if not self.Focused then return end
    if eventType == 5 then
        if context > 999 then
            local colidx = context - 1000
            local col = self:GetColumnAtPosition(colidx)
            if index == 0 then
                col:GoLeft()
            elseif index == 1 then
                col:GoRight()
            elseif index == 2 then
                col:GoUp()
            elseif index == 3 then
                col:GoDown()
            end
            return
        end

        if self.CurrentColumnIndex == context then
            if index == self:CurrentColumn():Index() then
                self:CurrentColumn():Select()
                return
            end
            self:CurrentColumn():CurrentItem():Selected(false)
            self:CurrentColumn():Index(index)
            self:CurrentColumn():CurrentItem():Selected(true)
            if self:CurrentColumn().type == PLT_COLUMNS.SETTINGS then
                AddTextEntry("PAUSEMENU_Current_Description", self:CurrentColumn():CurrentItem():Description())
            end
        else
            local selectedCol = self:GetColumnAtPosition(context)
            self:SwitchColumn(context)
            selectedCol:CurrentItem():Selected(false)
            selectedCol:Index(index)
            selectedCol:CurrentItem():Selected(true)
            if selectedCol.type == PLT_COLUMNS.SETTINGS then
                AddTextEntry("PAUSEMENU_Current_Description", selectedCol:CurrentItem():Description())
            end
        end
    elseif eventType == 10 or eventType == 11 then
        local dir = -1
        if eventType == 11 then
            dir = 1
        end
        self:CurrentColumn():MouseScroll(dir)
    end
end

function PlayerListTab:Populate()
    ScaleformUI.Scaleforms._pauseMenu._pause:CallFunction("SET_MENU_LEVEL", 1)
    self:StateChange()
    ScaleformUI.Scaleforms._pauseMenu._pause:CallFunction("SET_MENU_LEVEL", 0)
    if self.LeftColumn ~= nil then
        self.LeftColumn:Populate()
    end
    if self.CenterColumn ~= nil then
        self.CenterColumn:Populate()
    end
    if self.RightColumn ~= nil then
        self.RightColumn:Populate()
    end
end

function PlayerListTab:ShowColumns()
    if self.LeftColumn ~= nil then
        self.LeftColumn:ShowColumn()
    end
    if self.CenterColumn ~= nil then
        self.CenterColumn:ShowColumn()
    end
    if self.RightColumn ~= nil then
        self.RightColumn:ShowColumn()
    end
end

function PlayerListTab:Focus()
    BaseTab.Focus(self)
    self:CurrentColumn().Focused = true
    self:CurrentColumn():CurrentItem():Selected(true)
    if self:CurrentColumn().type == PLT_COLUMNS.SETTINGS then
        AddTextEntry("PAUSEMENU_Current_Description", self:CurrentColumn():CurrentItem():Description())
    elseif self:CurrentColumn().type == PLT_COLUMNS.PLAYERS then
        ClearPedInPauseMenu()
        if self:CurrentColumn().Panel ~= nil then
            self:CurrentColumn().Panel:UpdatePanel()
            if self.RightColumn ~= nil then
                self.RightColumn:ColumnVisible(false)
            end
        elseif self.RightColumn ~= nil then
            self.RightColumn:ColumnVisible(true)
        end
    end
    ScaleformUI.Scaleforms._pauseMenu._pause:CallFunction("SET_COLUMN_HIGHLIGHT", self:CurrentColumn().position,
        self:CurrentColumn():Index() - 1, true, false)
end

function PlayerListTab:UnFocus()
    if self.CurrentColumnIndex > 0 then
        self:SelectColumn(1)
    end
    BaseTab.UnFocus(self)
    if self.LeftColumn then
        self.LeftColumn:CurrentItem():Selected(false)
        self.LeftColumn.Focused = false
    end
    if self.CenterColumn then
        self.CenterColumn:CurrentItem():Selected(false)
        self.CenterColumn.Focused = false
    end
    if self.RightColumn then
        self.RightColumn:CurrentItem():Selected(false)
        if not self.RightColumn:ColumnVisible() then
            self.RightColumn:ColumnVisible(true)
        end
        self.RightColumn.Focused = false
    end
    AddTextEntry("PAUSEMENU_Current_Description", "")
end

File: ScaleformUI_Csharp/Menus/Pause Menus/PauseMenu/Tabs/PlayerListTab.cs
using CitizenFX.Core.Native;
using ScaleformUI.Elements;
using ScaleformUI.Menus;
using ScaleformUI.PauseMenus.Elements;
using ScaleformUI.PauseMenus.Elements.Columns;
using ScaleformUI.PauseMenus.Elements.Panels;

namespace ScaleformUI.PauseMenu
{
    public enum PLT_COLUMNS
    {
        SETTINGS,
        PLAYERS,
        MISSIONS,
        STORE,
        MISSION_DETAILS,
    }

    public delegate void ColumnFocusedEvent(PM_Column column, int index);
    public class PlayerListTab : BaseTab
    {
        public event ColumnFocusedEvent OnFocusChanged;
        public bool ForceFirstSelectionOnFocus { get; set; }
        private int[] order = new int[3];
        public PlayerListTab(string name, SColor color) : base(name, color)
        {
            _type = 2;
            _identifier = "Page_Multi";
            Minimap = new MinimapPanel(this) { HidePedBlip = true };
        }

        public void SetupLeftColumn(PM_Column column)
        {
            if (column is MissionDetailsPanel)
                throw new Exception("You cannot set the mission details column as the left column.");
            column.position = PM_COLUMNS.LEFT;
            LeftColumn = column;
            LeftColumn.Parent = this;
            order[0] = column.type;
        }
        public void SetupCenterColumn(PM_Column column)
        {
            if (column is MissionDetailsPanel)
                throw new Exception("You cannot set the mission details column as the center column.");
            column.position = PM_COLUMNS.MIDDLE;
            CenterColumn = column;
            CenterColumn.Parent = this;
            order[1] = column.type;
        }
        public void SetupRightColumn(PM_Column column)
        {
            column.position = PM_COLUMNS.RIGHT;
            RightColumn = column;
            RightColumn.Parent = this;
            order[2] = column.type;
        }

        public void SwitchColumn(int index)
        {
            SwitchColumn((PM_COLUMNS)index);
        }

        public void SwitchColumn(PM_COLUMNS index)
        {
            switchColumnInternal(index);
        }

        private void switchColumnInternal(PM_COLUMNS index)
        {
            // for now we won't allow selecting the mission details column
            if (index > PM_COLUMNS.RIGHT || order[(int)index] == (int)PLT_COLUMNS.MISSION_DETAILS) return;
            bool canHideShow = true;
            var col = GetColumnAtPosition(index);

            // handle if the column is null
            if (col == null)
            {
                if (index < PM_COLUMNS.RIGHT)
                {
                    switchColumnInternal(index + ((int)index < CurrentColumnIndex?-1:1));
                }
                return;
            }

            // we don't check for right column because
            // if right column is players then it means there's no player panel to be shown
            if (LeftColumn is PlayerListColumn plCol)
            {
                if (plCol.CurrentItem.KeepPanelVisible)
                {
                    canHideShow = false;
                }
            }
            else if (CenterColumn is PlayerListColumn cenPlCol)
            {
                if(cenPlCol.CurrentItem.KeepPanelVisible)
                {
                    canHideShow = false;
                }
            }
            if (canHideShow)
            {
                if (Parent != null && Parent.Visible)
                {
                    if (col is PlayerListColumn plc)
                    {
                        if (plc.CurrentItem.Panel != null)
                        {
                            RightColumn.ColumnVisible = false;
                        }
                    }
                    else
                    {
                        bool show = true;
                        var befcol = GetColumnAtPosition(index - 1);
                        var afterCol = GetColumnAtPosition(index + 1);
                        if (befcol is PlayerListColumn _plc && !_plc.CurrentItem.KeepPanelVisible)
                        {
                            _plc.CurrentItem.Dispose();
                            show = false;
                        }

                        if (afterCol is PlayerListColumn _plc2 && !_plc2.CurrentItem.KeepPanelVisible)
                        {
                            _plc2.CurrentItem.Dispose();
                            show = false;
                        }

                        var _col = GetColumnAtPosition(2);
                        _col.ColumnVisible = !show;
                    }
                }
            }
            else
            {
                if(RightColumn != null)
                    RightColumn.ColumnVisible = false;
            }

            if (index == PM_COLUMNS.RIGHT && !canHideShow) return;

            CurrentColumnIndex = (int)index;
            Main.PauseMenu._pause.CallFunction("SET_MENU_LEVEL", CurrentColumnIndex + 1);
            Main.PauseMenu._pause.CallFunction("MENU_SHIFT_DEPTH", 0, true, true);
            Parent.focusLevel = CurrentColumnIndex + 1;
            Main.PauseMenu._pause.CallFunction("SET_COLUMN_HIGHLIGHT", (int)col.position, col.Index, true, true);
            col.Items[col.Index].Selected = true;
        }

        public override void StateChange(int state)
        {
            Main.PauseMenu._pause.CallFunction("MENU_STATE", string.Join("", order));
        }

        public override void GoUp()
        {
            if (!Focused) return;
            var col = GetColumnAtPosition(CurrentColumnIndex);
            col.GoUp();
            col.SetColumnScroll(col.Index + 1, col.Items.Count, col.VisibleItems, string.Empty, col.Items.Count < col.VisibleItems);
        }

        public override void GoDown()
        {
            if (!Focused) return;
            var col = GetColumnAtPosition(CurrentColumnIndex);
            col.GoDown();
            col.SetColumnScroll(col.Index + 1, col.Items.Count, col.VisibleItems, string.Empty, col.Items.Count < col.VisibleItems);
        }

        public override void MouseEvent(int eventType, int context, int index)
        {
            if (!Focused) return;
            switch (eventType)
            {
                case 5: // mouse click
                    {
                        if (context > 999)
                        {
                            var columnIndex = context - 1000;
                            var col = GetColumnAtPosition(columnIndex);
                            switch (index)
                            {
                                case 0:
                                    col.GoLeft();
                                    break;
                                case 1:
                                    col.GoRight();
                                    break;
                                case 2:
                                    col.GoUp();
                                    break;
                                case 3:
                                    col.GoDown();
                                    break;
                            }
                            return;
                        }

                        var currentCol = GetColumnAtPosition(CurrentColumnIndex);


                        if (CurrentColumnIndex == context)
                        {
                            if (index == currentCol.Index)
                            {
                                currentCol.Select();
                                return;
                            }
                            switch (currentCol)
                            {
                                case SettingsListColumn set:
                                    set.CurrentItem.Selected = false;
                                    set.CurrentSelection = index;
                                    set.CurrentItem.Selected = true;
                                    AddTextEntry("PAUSEMENU_Current_Description", set.CurrentItem.Description);
                                    break;
                                case PlayerListColumn play:
                                    ClearPedInPauseMenu();
                                    play.CurrentItem.Selected = false;
                                    play.CurrentSelection = index;
                                    play.CurrentItem.Selected = true;
                                    play.CurrentItem.CreateClonedPed();
                                    break;
                                case MissionsListColumn miss:
                                    miss.CurrentItem.Selected = false;
                                    miss.CurrentSelection = index;
                                    miss.CurrentItem.Selected = true;
                                    break;
                                case StoreListColumn store:
                                    store.CurrentItem.Selected = false;
                                    store.CurrentSelection = index;
                                    store.CurrentItem.Selected = true;
                                    break;
                            }
                        }
                        else
                        {
                            var selectedCol = GetColumnAtPosition(context);
                            switchColumnInternal((PM_COLUMNS)context);
                            switch (selectedCol)
                            {
                                case SettingsListColumn set:
                                    set.CurrentItem.Selected = false;
                                    set.CurrentSelection = index;
                                    set.CurrentItem.Selected = true;
                                    AddTextEntry("PAUSEMENU_Current_Description", set.CurrentItem.Description);
                                    break;
                                case PlayerListColumn play:
                                    ClearPedInPauseMenu();
                                    play.CurrentItem.Selected = false;
                                    play.CurrentSelection = index;
                                    play.CurrentItem.Selected = true;
                                    play.CurrentItem.CreateClonedPed();
                                    break;
                                case MissionsListColumn miss:
                                    miss.CurrentItem.Selected = false;
                                    miss.CurrentSelection = index;
                                    miss.CurrentItem.Selected = true;
                                    break;
                                case StoreListColumn store:
                                    store.CurrentItem.Selected = false;
                                    store.CurrentSelection = index;
                                    store.CurrentItem.Selected = true;
                                    break;
                            }
                        }
                    }
                    break;
                case 8: // not hover
                    break;
                case 9: // mouse hover
                    break;
                case 10: // mouse wheel up
                case 11: // mouse wheel down
                    {
                        var currentCol = GetColumnAtPosition(CurrentColumnIndex);
                        currentCol.MouseScroll(eventType == 10 ? -1 : 1);
                    }
                    break;
            }
        }

        public override void GoLeft()
        {
            if (!Focused) return;
            var col = GetColumnAtPosition(CurrentColumnIndex);
            col.GoLeft();
        }

        public override void GoRight()
        {
            if (!Focused) return;
            var col = GetColumnAtPosition(CurrentColumnIndex);
            col.GoRight();
        }
        public override void Select()
        {
            if (!Focused) return;
            var col = GetColumnAtPosition(CurrentColumnIndex);
            col.Select();
        }

        public override void GoBack()
        {
            if (!Focused) return;
            if (CurrentColumnIndex > 0)
            {
                var col = GetColumnAtPosition(CurrentColumnIndex);
                if(col != null)
                    col.Items[col.Index].Selected = false;
                Main.PauseMenu._pause.CallFunction("SET_COLUMN_FOCUS", CurrentColumnIndex, false, false, false);
                switchColumnInternal((PM_COLUMNS)CurrentColumnIndex - 1);
                col = GetColumnAtPosition(CurrentColumnIndex);
                if (col == null && CurrentColumnIndex > 0)
                {
                    GoBack();
                    return;
                }
                col.Items[col.Index].Selected = true;
                if (col is PlayerListColumn plc)
                {
                    if (plc.CurrentItem.Panel != null)
                    {
                        plc.CurrentItem.CreateClonedPed();
                        var _col = GetColumnAtPosition(2);
                        _col.ColumnVisible = false;
                    }
                }
                else
                {
                    var _col = GetColumnAtPosition(2);
                    if(_col != null)
                        _col.ColumnVisible = true;
                }
            }
        }

        public override void Refresh(bool highlightOldIndex)
        {
        }

        public override void Populate()
        {
            Parent._pause._pause.CallFunction("SET_MENU_LEVEL", 1);
            StateChange(0);
            Parent._pause._pause.CallFunction("SET_MENU_LEVEL", 0);
            LeftColumn?.Populate();
            CenterColumn?.Populate();
            RightColumn?.Populate();
        }

        public override void ShowColumns()
        {
            LeftColumn?.ShowColumn();
            CenterColumn?.ShowColumn();
            RightColumn?.ShowColumn();
        }

        public override void SetDataSlot(PM_COLUMNS slot, int index)
        {
        }

        public override void UpdateSlot(PM_COLUMNS slot, int index)
        {
        }

        public override void Focus()
        {
            base.Focus();
            var col = GetColumnAtPosition(CurrentColumnIndex);
            switch (col)
            {
                case SettingsListColumn set:
                    set.Focused = true;
                    set.CurrentItem.Selected = true;
                    AddTextEntry("PAUSEMENU_Current_Description", set.CurrentItem.Description);
                    break;
                case PlayerListColumn play:
                    ClearPedInPauseMenu();
                    play.Focused = true;
                    play.CurrentItem.Selected = true;
                    play.CurrentItem.CreateClonedPed();
                    if(play.CurrentItem.Panel != null)
                    {
                        play.CurrentItem.Panel.UpdatePanel();
                        if(RightColumn != null)
                        RightColumn.ColumnVisible = false;
                    }
                    else if (RightColumn != null)
                        RightColumn.ColumnVisible = true;
                    break;
                case MissionsListColumn miss:
                    miss.Focused = true;
                    miss.CurrentItem.Selected = true;
                    break;
                case StoreListColumn store:
                    store.Focused = true;
                    store.CurrentItem.Selected = true;
                    break;
            }
            Main.PauseMenu._pause.CallFunction("SET_COLUMN_HIGHLIGHT", (int)col.position, col.Index, true, false);
        }

        public override void UnFocus()
        {
            base.UnFocus();
            ClearPedInPauseMenu();
            if (LeftColumn != null)
            {
                LeftColumn.Focused = false;
                LeftColumn.Items[LeftColumn.Index].Selected = false;
            }

            if (CenterColumn != null)
            {
                CenterColumn.Focused = false;
                CenterColumn.Items[CenterColumn.Index].Selected = false;
            }

            if (RightColumn != null)
            {
                RightColumn.Focused = false;
                RightColumn.Items[RightColumn.Index].Selected = false;
                if (!RightColumn.ColumnVisible)
                    RightColumn.ColumnVisible = true;
            }

            API.AddTextEntry("PAUSEMENU_Current_Description", "");
        }
    }
}